crazy_rl/multi_agent/jax/base_parallel_env.py

```diff
@@ -0,0 +1,389 @@
+        raise NotImplementedError
+
+    def _action_space(self) -> spaces.Space:
+        """Returns the action space of the environment. Must be implemented in a subclass."""
+        raise NotImplementedError
+
+    def _compute_obs(self, state):
+        """Returns the current observation of the environment. Must be implemented in a subclass."""
+        raise NotImplementedError
+
+    def _compute_action(self, actions, state):
+        """Computes the action passed to `.step()` into action matching the mode environment. Must be implemented in a subclass.
+
+        Args:
+            actions : ndarray | dict[..]. The input action for one drones
+            state : the state of the environment (contains agent_location used in this function)
+        """
+        raise NotImplementedError
+
+    def _compute_reward(self, state):
+        """Computes the current reward value(s). Must be implemented in a subclass."""
+        raise NotImplementedError
+
+    def _compute_terminated(self, state):
+        """Computes the current done value(s). Must be implemented in a subclass."""
+        raise NotImplementedError
+
+    def _compute_truncation(self, state):
+        """Computes the current done value(s). Must be implemented in a subclass."""
+        raise NotImplementedError
+
+    # PettingZoo API
+    @override
+    def reset(self, state, seed=None, return_info=False, options=None):
+        """
+        if self._mode == "real":
+            # self.swarm.parallel_safe(reset_estimator)
+            state = jdc.replace(state, agent_location=self._get_drones_state(self._mode, state.agent_location))
+            print("reset", state.agent_location)
+
+            command = dict()
+            # dict target_position URI
+            for agent in range(self.num_drones):
+                uri = "radio://0/4/2M/E7E7E7E7" + str(agent).zfill(2)
+                target = self._init_flying_pos[agent]
+                agent = state.agent_location[agent]
+                command[uri] = [[agent, target]]
+
+            self.swarm.parallel_safe(run_take_off)
+            print("Take off successful.")
+            print(f"Setting the drone positions to the initial positions. {command}")
+            self.swarm.parallel_safe(run_sequence, args_dict=command)
+
+            state = jdc.replace(state, agent_location=self._get_drones_state(self._mode, state.agent_location))
+
+        else:
+        """
+        state = jdc.replace(state, agent_location=jnp.copy(self._init_flying_pos))
+
+        state = jdc.replace(state, timestep=0)
+
+        state = self._compute_obs(state)
+
+        if self.render_mode == "human" and self._mode == "simu":
+            self._render_frame()
+
+        state = jdc.replace(state, terminations=jnp.zeros(self.num_drones), truncations=jnp.zeros(self.num_drones))
+
+        return state
+
+    @partial(jit, static_argnums=(0,))
+    def _compute_step(self, state):
```

Comment:
Can't we move the content of this to step directly? So that I can jit step overall